runtime/src/main/java/org/corfudb/util/Utils.java

```diff
@@ -205,99 +229,78 @@ public static void updateCommittedTail(RuntimeLayout runtimeLayout,
+        // Since a node can exist as a head for multiple segments we need to a set to
+        // coalesce the candidates to unique nodes only
+        Set<String> segmentsHeadNodes = getChainHeadFromAllSegments(runtimeLayout.getLayout());
 
-        Layout.LayoutSegment segment = runtimeLayout.getLayout().getLatestSegment();
-
-        // Query the tail of the head log unit in every stripe.
-        if (segment.getReplicationMode() == Layout.ReplicationMode.CHAIN_REPLICATION) {
-            for (Layout.LayoutStripe stripe : segment.getStripes()) {
-
-                TailsResponse res = CFUtils.getUninterruptibly(runtimeLayout
-                                .getLogUnitClient(stripe.getLogServers().get(DEFAULT_LOGUNIT))
-                                .getAllTails());
-                luResponses.add(res);
-            }
-        } else if (segment.getReplicationMode() == Layout.ReplicationMode.QUORUM_REPLICATION) {
-            throw new UnsupportedOperationException();
-        }
+        long[] globalTail = {Address.NON_EXIST};
+        final Map<UUID, Long> streamTails = new HashMap<>();
 
-        return aggregateLogUnitTails(luResponses);
-    }
-
-    /**
-     * Given a set of request tails, we aggregate them and maintain
-     * the greatest address per stream and the greatest tail over
-     * all responses.
-     * @param responses a set of tail responses
-     * @return An max-aggregation of all tails
-     */
-    static TailsResponse aggregateLogUnitTails(Set<TailsResponse> responses) {
-        long globalTail = Address.NON_ADDRESS;
-        Map<UUID, Long> globalStreamTails = new HashMap<>();
-
-        for (TailsResponse res : responses) {
-            globalTail = Math.max(globalTail, res.getLogTail());
-
-            for (Map.Entry<UUID, Long> stream : res.getStreamTails().entrySet()) {
-                long streamTail = globalStreamTails.getOrDefault(stream.getKey(), Address.NON_ADDRESS);
-                globalStreamTails.put(stream.getKey(), Math.max(streamTail, stream.getValue()));
-            }
-        }
-        // All epochs should be equal as all the tails are queried using a single runtime layout.
-        return new TailsResponse(globalTail, globalStreamTails);
+        segmentsHeadNodes.stream()
+                .map(node -> runtimeLayout
+                        .getLogUnitClient(node)
+                        .getAllTails())
+                .map(CFUtils::getUninterruptibly)
+                .forEach(resp -> {
+                    // All responses should be computed on the same epoch
+                    checkArgument(resp.getEpoch() == runtimeLayout.getLayout().getEpoch());
+                    // Find the global max global tail and stream tails across all responses
+                    globalTail[0] = Long.max(resp.getLogTail(), globalTail[0]);
+                    resp.getStreamTails().forEach((k, v) -> streamTails.merge(k, v, Long::max));
+                });
```

Comment:
Because of java stream's lazy evaluation property, if CFUtils.get is put in map(), then the futures are actually executed sequentially. We need to parallelize it to get better performance, by gathering all the futures first and then wait on them, similar to this: https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/util/Utils.java#L152-L162

Reply:
Yea, I thought about it, but didnt think it was worth it.
Will change it anyways.